File: backend/src/dearth/ModFilterService.ts
import { FileExtractor } from "./utils/FileExtractor.js";
import { FileOperator } from "./utils/FileOperator.js";
import { HashFilter } from "./strategies/HashFilter.js";
import { MixinFilter } from "./strategies/MixinFilter.js";
import { DexpubFilter } from "./strategies/DexpubFilter.js";
import { ModrinthFilter } from "./strategies/ModrinthFilter.js";
import { IFilterConfig, IFilterStrategy } from "./types.js";
import { logger } from "../utils/logger.js";
import { MessageWS } from "../utils/ws.js";
import path from "node:path";

export class ModFilterService {
  private readonly extractor: FileExtractor;
  private readonly operator: FileOperator;
  private readonly config: IFilterConfig;
  private messageWS?: MessageWS;
  private extraStrategies: IFilterStrategy[];

  constructor(modsPath: string, movePath: string, config: IFilterConfig, messageWS?: MessageWS, extraStrategies?: IFilterStrategy[]) {
    this.extractor = new FileExtractor(modsPath);
    this.operator = new FileOperator(movePath);
    this.config = config;
    this.messageWS = messageWS;
    this.extraStrategies = extraStrategies || [];
  }

  async filter(): Promise<void> {
    logger.info("开始模组筛选流程");
    const startTime = Date.now();

    try {
      const files = await this.extractor.extractFilesInfo();

      if (this.messageWS) {
        this.messageWS.filterModsStart(files.length);
      }

      const clientMods = await this.identifyClientSideMods(files);
      const result = await this.operator.moveClientSideMods(clientMods);

      const duration = Date.now() - startTime;

      if (this.messageWS) {
        this.messageWS.filterModsComplete(clientMods.length, result.success, duration);
      }

      logger.info("模组筛选流程完成", { 
        识别到的客户端模组: clientMods.length,
        成功移动: result.success,
        跳过: result.skipped,
        失败: result.error
      });
    } catch (error) {
      if (this.messageWS) {
        this.messageWS.filterModsError(error instanceof Error ? error.message : String(error));
      }
      throw error;
    }
  }

  private async identifyClientSideMods(files: Array<{ filename: string; hash: string; mixins: any[]; infos: any[] }>): Promise<string[]> {
    const clientMods: string[] = [];
    const processedFiles = new Set<string>();

    const useReplacement = this.extraStrategies.some(s => s.replacesBuiltin);

    if (!useReplacement) {
      const dexpubEnabled = this.config.dexpub;
      const mixinsEnabled = this.config.mixins;

      const [dexpubResult, mixinResult] = await Promise.all([
        dexpubEnabled ? this.runDexpubFilter(files) : Promise.resolve(null),
        mixinsEnabled ? this.runMixinFilter(files) : Promise.resolve(null),
      ]);

      let serverModsSet = new Set<string>();
      if (dexpubResult) {
        logger.info("Galaxy Square (dexpub) 检查完成", { 服务端模组: dexpubResult.serverMods, 客户端模组: dexpubResult.clientMods });
        dexpubResult.clientMods.forEach(mod => processedFiles.add(mod));
        serverModsSet = new Set(dexpubResult.serverMods);
        serverModsSet.forEach(mod => processedFiles.add(mod));
        clientMods.push(...dexpubResult.clientMods);

        if (this.messageWS) {
          this.messageWS.filterModsProgress(processedFiles.size, files.length, "Galaxy Square (dexpub) 检查");
        }
      }

      if (mixinResult) {
        logger.info("Mixin 检查完成", { 客户端模组数: mixinResult.length });
        mixinResult.forEach(mod => processedFiles.add(mod));
        clientMods.push(...mixinResult);

        if (this.messageWS) {
          this.messageWS.filterModsProgress(processedFiles.size, files.length, "Mixin 检查");
        }
      }

      if (this.config.modrinth) {
        logger.info("开始 Modrinth API 检查客户端模组");

        const unprocessedFiles = files.filter(f => !processedFiles.has(f.filename));
        const modrinthMods = await new ModrinthFilter().filter(unprocessedFiles);

        modrinthMods.forEach(mod => processedFiles.add(mod));
        clientMods.push(...modrinthMods);

        if (this.messageWS) {
          this.messageWS.filterModsProgress(processedFiles.size, files.length, "Modrinth API 检查");
        }
      }

      if (this.config.hashes) {
        logger.info("开始 Hash 检查客户端模组");

        const unprocessedFiles = files.filter(f => !processedFiles.has(f.filename));
        const hashMods = await new HashFilter().filter(unprocessedFiles);

        clientMods.push(...hashMods);

        if (this.messageWS) {
          this.messageWS.filterModsProgress(processedFiles.size, files.length, "Hash 检查");
        }
      }
    }

    for (const strategy of this.extraStrategies) {
      logger.info(`开始插件策略检查: ${strategy.name}`);
      const unprocessedFiles = files.filter(f => !processedFiles.has(f.filename));
      if (unprocessedFiles.length === 0) break;
      const mods = await strategy.filter(unprocessedFiles);
      mods.forEach((mod: string) => processedFiles.add(mod));
      clientMods.push(...mods);
      if (this.messageWS) {
        this.messageWS.filterModsProgress(processedFiles.size, files.length, strategy.name);
      }
    }

    const uniqueMods = [...new Set(clientMods)];
    logger.info("识别到客户端模组", { 数量: uniqueMods.length, 模组: uniqueMods });

    if (uniqueMods.length > 0) {
      logger.debug("第一个模组路径", { 原始路径: uniqueMods[0], 绝对路径: path.resolve(uniqueMods[0]), cwd: process.cwd() });
    }

    return uniqueMods;
  }

  private async runDexpubFilter(files: Array<{ filename: string; hash: string; mixins: any[]; infos: any[] }>): Promise<{ clientMods: string[]; serverMods: string[] }> {
    logger.info("开始 Galaxy Square (dexpub) 检查客户端模组");
    const dexpubStrategy = new DexpubFilter();
    const clientMods = await dexpubStrategy.filter(files);
    const serverMods = await dexpubStrategy.getServerMods(files);
    return { clientMods, serverMods };
  }

  private async runMixinFilter(files: Array<{ filename: string; hash: string; mixins: any[]; infos: any[] }>): Promise<string[]> {
    logger.info("开始 Mixin 检查客户端模组");
    return new MixinFilter().filter(files);
  }
}

File: backend-net/src/DeEarthX.Dearth/ModFilterService.cs
using System.Diagnostics;
using DeEarthX.Core.Abstractions;
using DeEarthX.Core.Configuration;
using DeEarthX.Core.Filter;
using DeEarthX.Realtime;

namespace DeEarthX.Dearth;

public sealed class ModFilterService
{
    private readonly IAppDirectoryProvider _appDirectory;
    private readonly ILogService _log;
    private readonly FileExtractor _extractor;
    private readonly FileOperator _operator;
    private readonly HashFilter _hashFilter;
    private readonly DexpubFilter _dexpubFilter;
    private readonly MixinFilter _mixinFilter;
    private readonly ModrinthFilter _modrinthFilter;

    public ModFilterService(
        IAppDirectoryProvider appDirectory,
        ILogService log,
        FileExtractor extractor,
        FileOperator fileOperator,
        HashFilter hashFilter,
        DexpubFilter dexpubFilter,
        MixinFilter mixinFilter,
        ModrinthFilter modrinthFilter)
    {
        _appDirectory = appDirectory;
        _log = log;
        _extractor = extractor;
        _operator = fileOperator;
        _hashFilter = hashFilter;
        _dexpubFilter = dexpubFilter;
        _mixinFilter = mixinFilter;
        _modrinthFilter = modrinthFilter;
    }

    public async Task<FilterResult> FilterAsync(
        string modsDir,
        string modpackName,
        DeEarthXConfig config,
        IEnumerable<IFilterStrategy>? extraStrategies,
        IMessageService? message,
        CancellationToken ct = default)
    {
        var startTime = Stopwatch.StartNew();
        var filterConfig = new DeEarthXFilterConfig(config);

        try
        {
            _log.Info("开始模组筛选流程");
            var files = await _extractor.ExtractFilesInfoAsync(modsDir, ct).ConfigureAwait(false);

            if (message is not null)
            {
                await message.FilterModsStart(files.Count).ConfigureAwait(false);
            }

            var clientMods = new HashSet<string>(StringComparer.Ordinal);
            var processedFiles = new HashSet<string>(StringComparer.Ordinal);

            const bool useReplacement = false;

            if (!useReplacement)
            {
                var dexpubTask = filterConfig.Dexpub
                    ? _dexpubFilter.CheckAsync(files)
                    : Task.FromResult<(HashSet<string> Client, HashSet<string> Server)>((new(StringComparer.Ordinal), new(StringComparer.Ordinal)));
                var mixinTask = filterConfig.Mixins
                    ? _mixinFilter.FilterBatchAsync(files)
                    : Task.FromResult(new HashSet<string>(StringComparer.Ordinal));

                await Task.WhenAll(dexpubTask, mixinTask).ConfigureAwait(false);
                var dexpubResult = await dexpubTask.ConfigureAwait(false);
                var mixinResult = await mixinTask.ConfigureAwait(false);

                if (filterConfig.Dexpub)
                {
                    _log.Info("Galaxy Square (dexpub) 检查完成", new { client = dexpubResult.Client.Count, server = dexpubResult.Server.Count });
                    foreach (var m in dexpubResult.Client)
                    {
                        processedFiles.Add(m);
                        clientMods.Add(m);
                    }

                    if (message is not null)
                    {
                        await message.FilterModsProgress(processedFiles.Count, files.Count, "Galaxy Square (dexpub) 检查").ConfigureAwait(false);
                    }
                }

                if (filterConfig.Mixins)
                {
                    _log.Info("Mixin 检查完成", new { count = mixinResult.Count });
                    foreach (var m in mixinResult)
                    {
                        processedFiles.Add(m);
                        clientMods.Add(m);
                    }

                    if (message is not null)
                    {
                        await message.FilterModsProgress(processedFiles.Count, files.Count, "Mixin 检查").ConfigureAwait(false);
                    }
                }

                if (filterConfig.Modrinth)
                {
                    _log.Info("开始 Modrinth API 检查客户端模组");
                    var unprocessed = files.Where(f => !processedFiles.Contains(f.FilePath)).ToList();
                    var modrinthMods = await _modrinthFilter.FilterBatchAsync(unprocessed).ConfigureAwait(false);
                    foreach (var m in modrinthMods)
                    {
                        processedFiles.Add(m);
                        clientMods.Add(m);
                    }

                    if (message is not null)
                    {
                        await message.FilterModsProgress(processedFiles.Count, files.Count, "Modrinth API 检查").ConfigureAwait(false);
                    }
                }

                if (filterConfig.Hashes)
                {
                    _log.Info("开始 Hash 检查客户端模组");
                    var unprocessed = files.Where(f => !processedFiles.Contains(f.FilePath)).ToList();
                    var hashMods = await _hashFilter.FilterBatchAsync(unprocessed).ConfigureAwait(false);
                    foreach (var m in hashMods)
                    {
                        clientMods.Add(m);
                    }

                    if (message is not null)
                    {
                        await message.FilterModsProgress(processedFiles.Count, files.Count, "Hash 检查").ConfigureAwait(false);
                    }
                }
            }

            if (extraStrategies is not null)
            {
                foreach (var strategy in extraStrategies)
                {
                    _log.Info($"开始插件策略检查: {strategy.Name}");
                    var unprocessed = files.Where(f => !processedFiles.Contains(f.FilePath)).ToList();
                    if (unprocessed.Count == 0)
                    {
                        break;
                    }

                    var adapter = new PluginStrategyAdapter(strategy, _extractor);
                    var mods = await adapter.FilterBatchAsync(unprocessed).ConfigureAwait(false);
                    foreach (var m in mods)
                    {
                        processedFiles.Add(m);
                        clientMods.Add(m);
                    }

                    if (message is not null)
                    {
                        await message.FilterModsProgress(processedFiles.Count, files.Count, strategy.Name).ConfigureAwait(false);
                    }
                }
            }

            var uniqueMods = clientMods.ToList();
            _log.Info("识别到客户端模组", new { count = uniqueMods.Count });

            var moveDir = Path.Combine(_appDirectory.GetAppDirectory(), ".rubbish", modpackName);
            var (success, error, skipped) = await _operator.MoveFilesAsync(uniqueMods, moveDir, ct).ConfigureAwait(false);

            startTime.Stop();
            var duration = startTime.ElapsedMilliseconds;

            if (message is not null)
            {
                await message.FilterModsComplete(uniqueMods.Count, success, duration).ConfigureAwait(false);
            }

            _log.Info("模组筛选流程完成", new
            {
                filtered = uniqueMods.Count,
                moved = success,
                skipped,
                error
            });

            return new FilterResult(uniqueMods.Count, success, skipped, error);
        }
        catch (Exception ex)
        {
            if (message is not null)
            {
                await message.FilterModsError(ex.Message).ConfigureAwait(false);
            }

            throw;
        }
    }
}
